# Zadanie rekrutacyjne dla firmy Novomatic
## Autor: Radosław Mocarski

### Dane:
Ze względu na liczne błędy w wersji `.csv` (widoczne m.in. w kolumnie `Supported languages` oraz w dyskusjach na stronie Kaggle), zdecydowałem się na użycie danych w formacie `.json`, które są rekomendowane przez społeczność, zamiast wersji `.csv` sugerowanej w poleceniu.

## Eksploracja danych:

### Wstępne wczytanie i ogarnięcie

Format danych jest dla mnie zbyt brutalny (mam tylko 16gb ramu w laptopie), więc stworzyłem skrypt `transform_json_to_parquet.py` który bierze pandasem plik json i konwertuje na wydajniejszy format parquet. Próbuje przy okazji rzutować na odpowiednie typy danych (boole, inty, floaty, daty i stringi). Niestety zmieniana jest informacja o listach i nullach w polach tekstowych, ale to opiszę później. Spójrzmy w dane:

In [1]:
import polars as pl
parquet_path = './archive/games.parquet'

print("Wczytywanie pliku Parquet...")
df = pl.read_parquet(parquet_path)

print("\nPodstawowe informacje o zbiorze danych (Schema):")
schema = df.schema
print(schema)

print("\nStatystyki dla kolumn:")
display(df.describe())

Wczytywanie pliku Parquet...

Podstawowe informacje o zbiorze danych (Schema):
Schema({'name': String, 'release_date': Datetime(time_unit='ns', time_zone=None), 'required_age': Int64, 'price': Float64, 'dlc_count': Int64, 'detailed_description': String, 'about_the_game': String, 'short_description': String, 'reviews': String, 'header_image': String, 'website': String, 'support_url': String, 'support_email': String, 'windows': Boolean, 'mac': Boolean, 'linux': Boolean, 'metacritic_score': Int64, 'metacritic_url': String, 'achievements': Int64, 'recommendations': Int64, 'notes': String, 'supported_languages': String, 'full_audio_languages': String, 'packages': String, 'developers': String, 'publishers': String, 'categories': String, 'genres': String, 'screenshots': String, 'movies': Float64, 'user_score': Int64, 'score_rank': Float64, 'positive': Int64, 'negative': Int64, 'estimated_owners': String, 'average_playtime_forever': Int64, 'average_playtime_2weeks': Int64, 'median_playtime_for

statistic,name,release_date,required_age,price,dlc_count,detailed_description,about_the_game,short_description,reviews,header_image,website,support_url,support_email,windows,mac,linux,metacritic_score,metacritic_url,achievements,recommendations,notes,supported_languages,full_audio_languages,packages,developers,publishers,categories,genres,screenshots,movies,user_score,score_rank,positive,negative,estimated_owners,average_playtime_forever,average_playtime_2weeks,median_playtime_forever,median_playtime_2weeks,discount,peak_ccu,tags,__index_level_0__
str,str,str,f64,f64,f64,str,str,str,str,str,str,str,str,f64,f64,f64,f64,str,f64,f64,str,str,str,str,str,str,str,str,str,f64,f64,f64,f64,f64,str,f64,f64,f64,f64,f64,f64,str,f64
"""count""","""122611""","""122611""",122611.0,122611.0,122611.0,"""122611""","""122611""","""122611""","""122611""","""122611""","""122611""","""122611""","""122611""",122611.0,122611.0,122611.0,122611.0,"""122611""",122611.0,122611.0,"""122611""","""122611""","""122611""","""122611""","""122611""","""122611""","""122611""","""122611""","""122611""",0.0,122611.0,40.0,122611.0,122611.0,"""122611""",122611.0,122611.0,122611.0,122611.0,122611.0,122611.0,"""122611""",122611.0
"""null_count""","""0""","""0""",0.0,0.0,0.0,"""0""","""0""","""0""","""0""","""0""","""0""","""0""","""0""",0.0,0.0,0.0,0.0,"""0""",0.0,0.0,"""0""","""0""","""0""","""0""","""0""","""0""","""0""","""0""","""0""",122611.0,0.0,122571.0,0.0,0.0,"""0""",0.0,0.0,0.0,0.0,0.0,0.0,"""0""",0.0
"""mean""",null,"""2022-01-21 08:27:37.847990""",0.167611,4.765091,0.545856,null,null,null,null,null,null,null,null,0.999641,0.173655,0.128096,2.564941,null,18.087015,961.824967,null,null,null,null,null,null,null,null,null,null,0.024549,99.175,1044.986135,169.197429,null,208.02322,13.789268,173.570528,14.72217,18.353663,54.593324,null,1.9854e6
"""std""",null,null,1.653591,12.53103,14.516026,null,null,null,null,null,null,null,null,null,null,null,13.660559,null,141.493879,21878.803992,null,null,null,null,null,null,null,null,null,null,1.394901,0.675107,28091.729937,5374.64488,null,11217.684738,270.378053,11202.537481,294.509615,28.85897,3729.451567,null,1.0876e6
"""min""","""""","""1997-06-30 00:00:00""",0.0,0.0,0.0,"""""","""""","""""","""""","""""","""""","""""","""""",0.0,0.0,0.0,0.0,"""""",0.0,0.0,"""""","""['Arabic', 'English']""","""['Arabic', 'English']""","""[]""","""[""'What Day is it?' Games""]""","""[""1mPaCt' (JSD)"", 'DasProfGame…","""['Adjustable Text Size', 'Colo…","""['Accounting', 'Animation & Mo…","""['https://shared.akamai.steams…",null,0.0,98.0,0.0,0.0,"""0 - 0""",0.0,0.0,0.0,0.0,0.0,0.0,"""[]""",10.0
"""25%""",null,"""2019-12-20 00:00:00""",0.0,0.55,0.0,null,null,null,null,null,null,null,null,null,null,null,0.0,null,0.0,0.0,null,null,null,null,null,null,null,null,null,null,0.0,99.0,0.0,0.0,null,0.0,0.0,0.0,0.0,0.0,0.0,null,1.06319e6
"""50%""",null,"""2022-11-15 00:00:00""",0.0,2.24,0.0,null,null,null,null,null,null,null,null,null,null,null,0.0,null,2.0,0.0,null,null,null,null,null,null,null,null,null,null,0.0,99.0,5.0,1.0,null,0.0,0.0,0.0,0.0,0.0,0.0,null,1.90738e6
"""75%""",null,"""2024-10-04 00:00:00""",0.0,5.24,0.0,null,null,null,null,null,null,null,null,null,null,null,0.0,null,19.0,0.0,null,null,null,null,null,null,null,null,null,null,0.0,100.0,37.0,10.0,null,0.0,0.0,0.0,0.0,40.0,0.0,null,2.86959e6
"""max""","""𣸩""","""2026-01-05 00:00:00""",21.0,999.98,3703.0,"""🫧 A Whimsical Bubble-Popping A…","""🫧 A Whimsical Bubble-Popping A…","""🩸🌌💀🌫️⛪️""","""게임 분위기 – 어둡고 무서운 사운드로 인해 긴장감을 …","""https://shared.akamai.steamsta…","""www.windybeard.com""","""暂无""","""𝗰𝗵𝗶𝗰𝗮𝗴𝗼𝗽𝗿𝗼𝗷𝗲𝗰𝘁@𝟭𝟲𝟯.𝗰𝗼𝗺""",1.0,1.0,1.0,97.0,"""https://www.metacritic.com/gam…",9821.0,4.830455e6,"""🪀 CONTENT WARNINGS 🪀 TOY BOX c…","""[]""","""[]""","""[{'title': 'Yearly Subscriptio…","""[]""","""[]""","""[]""","""[]""","""[]""",null,100.0,100.0,7.642084e6,1.173003e6,"""50000000 - 100000000""",3.429544e6,20088.0,3.429544e6,20088.0,100.0,1.013936e6,"""{'e-sports': 220, 'FPS': 215, …",4.2643

In [2]:
pl.Config.set_tbl_cols(-1)
pl.Config.set_tbl_width_chars(1000)
print(df.head(3))

shape: (3, 43)
┌─────────────────────────────────┬─────────────────────┬──────────────┬───────┬───────────┬─────────────────────────────────┬─────────────────────────────────┬─────────────────────────────────┬─────────┬─────────────────────────────────┬─────────────────────────────────┬─────────────────────────────────┬────────────────────────┬─────────┬───────┬───────┬──────────────────┬────────────────┬──────────────┬─────────────────┬───────┬─────────────────────────────────┬──────────────────────┬─────────────────────────────────┬─────────────────┬────────────────┬─────────────────────────────────┬───────────────┬─────────────────────────────────┬────────┬────────────┬────────────┬──────────┬──────────┬──────────────────┬──────────────────────────┬─────────────────────────┬─────────────────────────┬────────────────────────┬──────────┬──────────┬─────────────────────────────────┬───────────────────┐
│ name                            ┆ release_date        ┆ required_age ┆ price ┆ dlc

Niestety można zauważyć, że w polach tekstowych jest masa elementów pustych (o treści `""` lub `"[]"`) i nie ma żadnej pustej, więc wypada je zmienić na wartości `null`. Same listy zostały potraktowane jako tekst, a jego elementy są w klamrach, więc i to wypada zmienić.

In [3]:
list_columns = [
    'supported_languages', 'full_audio_languages', 'developers', 
    'publishers', 'categories', 'genres', 'screenshots'
]

needs_cleaning = False
if 'categories' in schema and schema['categories'] == pl.String:
    needs_cleaning = True

if needs_cleaning:
    string_columns = [col for col, dtype in schema.items() if dtype == pl.String]
    null_cleaning_exprs = []
    
    for col_name in string_columns:
        null_cleaning_exprs.append(
            pl.when(pl.col(col_name).str.strip_chars().is_in(["", "[]", "{}", "null", "None", "NaN"]))
            .then(None)
            .otherwise(pl.col(col_name))
            .alias(col_name)
        )
    
    df = df.with_columns(null_cleaning_exprs)
    
    list_conversion_exprs = []
    for col_name in list_columns:
        if col_name in schema:
            list_conversion_exprs.append(
                pl.col(col_name)
                .str.strip_chars("[]") 
                .str.split(", ")       
                .list.eval(            
                    pl.element().str.strip_chars("'\" ") 
                )
                .alias(col_name)
            )
            
    df = df.with_columns(list_conversion_exprs)
    df.write_parquet(parquet_path)

else:
    print("Plik jest już wyczyszczony! Posiada poprawne listy i nulle. Pomijam czyszczenie.")

print("\nObecny schemat (wyrywkowe kolumny):")
preview_columns = ['categories', 'developers', 'supported_languages', 'tags']
for col_name in preview_columns:
    if col_name in df.schema:
        print(f" - {col_name}: {df.schema[col_name]}")


Obecny schemat (wyrywkowe kolumny):
 - categories: List(String)
 - developers: List(String)
 - supported_languages: List(String)
 - tags: String


In [4]:
print("\nStatystyki dla kolumn:")
display(df.describe())


Statystyki dla kolumn:


statistic,name,release_date,required_age,price,dlc_count,detailed_description,about_the_game,short_description,reviews,header_image,website,support_url,support_email,windows,mac,linux,metacritic_score,metacritic_url,achievements,recommendations,notes,supported_languages,full_audio_languages,packages,developers,publishers,categories,genres,screenshots,movies,user_score,score_rank,positive,negative,estimated_owners,average_playtime_forever,average_playtime_2weeks,median_playtime_forever,median_playtime_2weeks,discount,peak_ccu,tags,__index_level_0__
str,str,str,f64,f64,f64,str,str,str,str,str,str,str,str,f64,f64,f64,f64,str,f64,f64,str,f64,f64,str,f64,f64,f64,f64,f64,f64,f64,f64,f64,f64,str,f64,f64,f64,f64,f64,f64,str,f64
"""count""","""122610""","""122611""",122611.0,122611.0,122611.0,"""114191""","""114162""","""114302""","""12070""","""122530""","""49676""","""54192""","""100364""",122611.0,122611.0,122611.0,122611.0,"""4256""",122611.0,122611.0,"""22458""",114231.0,49881.0,"""96799""",114180.0,113784.0,113658.0,114198.0,116593.0,0.0,122611.0,40.0,122611.0,122611.0,"""122611""",122611.0,122611.0,122611.0,122611.0,122611.0,122611.0,"""83346""",122611.0
"""null_count""","""1""","""0""",0.0,0.0,0.0,"""8420""","""8449""","""8309""","""110541""","""81""","""72935""","""68419""","""22247""",0.0,0.0,0.0,0.0,"""118355""",0.0,0.0,"""100153""",8380.0,72730.0,"""25812""",8431.0,8827.0,8953.0,8413.0,6018.0,122611.0,0.0,122571.0,0.0,0.0,"""0""",0.0,0.0,0.0,0.0,0.0,0.0,"""39265""",0.0
"""mean""",null,"""2022-01-21 08:27:37.847990""",0.167611,4.765091,0.545856,null,null,null,null,null,null,null,null,0.999641,0.173655,0.128096,2.564941,null,18.087015,961.824967,null,null,null,null,null,null,null,null,null,null,0.024549,99.175,1044.986135,169.197429,null,208.02322,13.789268,173.570528,14.72217,18.353663,54.593324,null,1.9854e6
"""std""",null,null,1.653591,12.53103,14.516026,null,null,null,null,null,null,null,null,null,null,null,13.660559,null,141.493879,21878.803992,null,null,null,null,null,null,null,null,null,null,1.394901,0.675107,28091.729937,5374.64488,null,11217.684738,270.378053,11202.537481,294.509615,28.85897,3729.451567,null,1.0876e6
"""min""","""! Shakabula *""","""1997-06-30 00:00:00""",0.0,0.0,0.0,"""! ! ! PLEASE READ BEFORE BUYIN…","""! Important: 'X' key for selec…","""Hungry piggy vs Chicken: Yes…","""""ACDSee 15 is a full-featured …","""https://shared.akamai.steamsta…","""GranVitreous.com""","""*""","""*""",0.0,0.0,0.0,0.0,"""http://www.metacritic.com/game…",0.0,0.0,"""!! Important Notice !! This ga…",null,null,"""[{'title': ""Buy '1st Core: The…",null,null,null,null,null,null,0.0,98.0,0.0,0.0,"""0 - 0""",0.0,0.0,0.0,0.0,0.0,0.0,"""{""1990's"": 119, 'MMORPG': 112,…",10.0
"""25%""",null,"""2019-12-20 00:00:00""",0.0,0.55,0.0,null,null,null,null,null,null,null,null,null,null,null,0.0,null,0.0,0.0,null,null,null,null,null,null,null,null,null,null,0.0,99.0,0.0,0.0,null,0.0,0.0,0.0,0.0,0.0,0.0,null,1.06319e6
"""50%""",null,"""2022-11-15 00:00:00""",0.0,2.24,0.0,null,null,null,null,null,null,null,null,null,null,null,0.0,null,2.0,0.0,null,null,null,null,null,null,null,null,null,null,0.0,99.0,5.0,1.0,null,0.0,0.0,0.0,0.0,0.0,0.0,null,1.90738e6
"""75%""",null,"""2024-10-04 00:00:00""",0.0,5.24,0.0,null,null,null,null,null,null,null,null,null,null,null,0.0,null,19.0,0.0,null,null,null,null,null,null,null,null,null,null,0.0,100.0,37.0,10.0,null,0.0,0.0,0.0,0.0,40.0,0.0,null,2.86959e6
"""max""","""𣸩""","""2026-01-05 00:00:00""",21.0,999.98,3703.0,"""🫧 A Whimsical Bubble-Popping A…","""🫧 A Whimsical Bubble-Popping A…","""🩸🌌💀🌫️⛪️""","""게임 분위기 – 어둡고 무서운 사운드로 인해 긴장감을 …","""https://shared.akamai.steamsta…","""www.windybeard.com""","""暂无""","""𝗰𝗵𝗶𝗰𝗮𝗴𝗼𝗽𝗿𝗼𝗷𝗲𝗰𝘁@𝟭𝟲𝟯.𝗰𝗼𝗺""",1.0,1.0,1.0,97.0,"""https://www.metacritic.com/gam…",9821.0,4.830455e6,"""🪀 CONTENT WARNINGS 🪀 TOY BOX c…",null,null,"""[{'title': 'Yearly Subscriptio…",null,null,null,null,null,null,100.0,100.0,7.642084e6,1.173003e6,"""50000000 - 100000000""",3.429544e6,20088.0,3.429544e6,2008

In [5]:
print(df.head(3))

shape: (3, 43)
┌─────────────────────────────────┬─────────────────────┬──────────────┬───────┬───────────┬─────────────────────────────────┬─────────────────────────────────┬─────────────────────────────────┬─────────┬─────────────────────────────────┬─────────────────────────────────┬─────────────────────────────────┬────────────────────────┬─────────┬───────┬───────┬──────────────────┬────────────────┬──────────────┬─────────────────┬───────┬─────────────────────────────────┬──────────────────────┬─────────────────────────────────┬─────────────────┬────────────────┬─────────────────────────────────┬───────────────┬─────────────────────────────────┬────────┬────────────┬────────────┬──────────┬──────────┬──────────────────┬──────────────────────────┬─────────────────────────┬─────────────────────────┬────────────────────────┬──────────┬──────────┬─────────────────────────────────┬───────────────────┐
│ name                            ┆ release_date        ┆ required_age ┆ price ┆ dlc

Jak widać, nulle się pojawiły i to nawet niekiedy liczne. Nie zmieniłem do końca dwóch kolumn, mianowicie `tags` i `packages` przez ich ciężką późniejszą analizę. 

### Kolumny Tags i Packages

Tagi są słownikiem z popularnością kategorii gry nadawaną przez użytkowników - im większa liczba przy tagu, tym więcej użytkowników zaklasyfikowało tą grę pod ten tag. 

Paczki są dużo bardziej złożonym tematem, ponieważ są to oferty handlowe różnych pakietów. Np gry z jakąś dodatkową zawartością. Jest to niezwykle trudna kolumna do rozpakowania, ponieważ jest to fragment kodu JSONa zrzucony na string... Dlatego zagłębię się w te dane głębiej. Najpierw wyświetlę jeden rekord, a potem ten najdłuższy z zestawu danych.

In [6]:
print(df.get_column("packages")[13])

[{'title': 'Buy Forests, Fields and Fortresses', 'description': '', 'subs': [{'text': 'Forests, Fields and Fortresses - $4.99 $1.49', 'description': '', 'price': 1.49}]}]


In [7]:
longest_text = (
    df.select(pl.col("packages"))
    .filter(pl.col("packages").is_not_null())
    .with_columns(leng=pl.col("packages").str.len_chars())
    .sort("leng", descending=True)
    .head(1)
    .get_column("packages")[0]
)

print(f"Długość tekstu: {len(longest_text)} znaków")
print(longest_text)

Długość tekstu: 2205 znaków
[{'title': 'Buy Battlefield™ 6', 'description': '', 'subs': [{'text': 'Battlefield™ 6 - $69.99 $48.99', 'description': '', 'price': 48.99}, {'text': 'Battlefield™ 6 Phantom Edition - $99.99 $79.99', 'description': '', 'price': 79.99}]}, {'title': 'Buy Battlefield Coins', 'description': 'Battlefield Coins can be used in the in-game Store to purchase Soldiers, Weapon Packages, Vehicle Skins, and a host of other cosmetic accessories and boosters. Can also be used to purchase the Battle Pass. Click HERE for more information.', 'subs': [{'text': '500 BFC for Battlefield™ 6 and REDSEC - $4.99', 'description': 'Battlefield Coins can be used in the in-game Store to purchase Soldiers, Weapon Packages, Vehicle Skins, and a host of other cosmetic accessories and boosters. Can also be used to purchase the Battle Pass. Click HERE for more information.', 'price': 4.99}, {'text': '1100 BFC for Battlefield™ 6 and REDSEC - $9.99', 'description': 'Battlefield Coins can be use

Informacje z tej kolumny wydają się dosyć cenne. Jeżeli twórcy gry zależy na robieniu licznych pakietów, to może mu nieco bardziej zależeć na tej grze, przez co np. bardziej ją dopracował czy też zwiększył nakłady na marketing. Jednocześnie może wskazywać na nachalność mikrotransakcji, podejściu do ucinania zawartości i sprzedaży w dodatkach itp. Szkoda byłoby zatracić tą informację, dlatego spróbuję ją jakoś później przetworzyć.

Chyba najracjonalniejszym podejściem będzie analiza tekstu w poszukiwaniu konkretnego ciągu znaków. Obserwuję tutaj dwie interesujące opcje:
- `{'title': `
- `, 'price': `

Są na tyle unikatowe, że nie powinny znajdować się w opisach gier, a jednocześnie jednoznacznie definiują pojedynczy niepowtarzalny element. Na przykładzie gry Battlefield wylistowanym powyżej, można stwierdzić, że wyszukiwanie pierwszego ciągu będzie definiowało liczbę wersji gry, natomiast drugiego powiększy dodatkowo o mini dodatki do gry jak np. płatna waluta w grze. 

Przyznam szczerze, że nie wiem czy oddzielanie tych informacji ma jakiś większy sens. Obie te informacje wskazują na podobny aspekt i raczej głównym elementem będzie informacja, czy twórca robi więcej niż jedną ofertę. A że zazwyczaj staram się minimalizować liczbę parametrów branych do modeli, to wydaje mi się, że zastosowanie selekcji po drugim ciągu będzie najlepsze (ponieważ zawiera jednocześnie informację o pierwszym).

Kolumna tagów jest dużo prostsza. Przez to że jest słownikiem zrzutowanym do tekstu, można łatwo wydobyć potrzebne informacje. Niewątpliwie kategorie gry (jakimi są tagi) taką informacją są. Podejść do zmiany tych danych na przystępne do analizy danych jest kilka. Jedną z popularnych metod jest stworzenie kolumn z każdą kategorią i jako one-hot encoding zaznaczać czy do niej należy czy też nie. Zbliżoną wersją jest zamiast stosowania booli użyć np int64 i przepisac te dane ze słownika tworząc macierz rzadką (sparse matrix). 

Problemem w tych podejściach jest to, że dataframe staje się dla użytkownika małoczytelny przez liczbę kolumn (choć te obecnie już są małoczytelne przez ich liczbę). Zachowuje się wtedy pełną informację o danych, lub znaczną ich część (w przypadku podejścia boolowego). Z tego też powodu, często stosuje się podejście wypisania jednej (lub kilku) najważniejszych kategorii, zostawienie ich w dataframe w formie tekstowej (kategorycznej), a później zastosowanie odpowiedniej biblioteki, która zrobi operacje dzielenia na boole tuż przed uczeniem maszynowym. Dzięki temu my widzimy czytelne dane, a komputer ma dla niego wygodniejszą formę danych.

Dlatego, żeby sprawdzić jak dużym problemem będzie podział na kategorie z intami/boolami w dataframe, zobaczmy ile unikatowych danych by było.

In [ ]:
df_tags = df.select(
    pl.col("tags")
    .str.extract_all(r"['\"]([^'\"]+)['\"]:")
    .list.eval(pl.element().str.strip_chars("'\""))
    .alias("tag_list")
)

all_tags_count = df_tags.select(
    pl.col("tag_list").explode().drop_nulls().n_unique()
).item()

top1_tag_count = df_tags.select(
    pl.col("tag_list").list.first().drop_nulls().n_unique()
).item()

top2_tag_count = df_tags.select(
    pl.col("tag_list").list.slice(1, 1).list.first().drop_nulls().n_unique()
).item()

print(f"1. Laczna liczba wszystkich unikalnych tagow na Steam: {all_tags_count}")
print(f"2. Liczba unikalnych tagow, jesli wezmiemy tylko 1. miejsce: {top1_tag_count}")
print(f"3. Liczba unikalnych tagow, jesli wezmiemy tylko 2. miejsce: {top2_tag_count}")

dimensionality_reduction = (1 - (top1_tag_count / all_tags_count)) * 100
print(f"\nWNIOSEK: Biorac tylko glowny tag, redukujemy wymiarowosc danych o ok. {dimensionality_reduction:.1f}%.")

1. Laczna liczba wszystkich unikalnych tagow na Steam: 452
2. Liczba unikalnych tagow, jesli wezmiemy tylko 1. miejsce: 419
3. Liczba unikalnych tagow, jesli wezmiemy tylko 2. miejsce: 434

WNIOSEK: Biorac tylko glowny tag, redukujemy wymiarowosc danych o ok. 7.3%.


Szybka weryfikacja tagów po zastosowanym regexie

In [19]:
print(df_tags.select(
    pl.col("tag_list").explode().drop_nulls().unique()
).head(10))

shape: (10, 1)
┌───────────────────────┐
│ tag_list              │
│ ---                   │
│ str                   │
╞═══════════════════════╡
│ TrackIR':             │
│ BMX':                 │
│ Football (American)': │
│ Spaceships':          │
│ Musou':               │
│ Resource Management': │
│ First-Person':        │
│ Idler':               │
│ Martial Arts':        │
│ Volleyball':          │
└───────────────────────┘


In [20]:
print(df_tags.select(
    pl.col("tag_list").list.slice(1, 1).list.first().drop_nulls().unique()
).head(10))

shape: (10, 1)
┌───────────────────────┐
│ tag_list              │
│ ---                   │
│ str                   │
╞═══════════════════════╡
│ Games Workshop':      │
│ Dungeon Crawler':     │
│ Bikes':               │
│ Funny':               │
│ Level Editor':        │
│ Based On A Novel':    │
│ Stylized':            │
│ Interactive Fiction': │
│ Comedy':              │
│ Space Sim':           │
└───────────────────────┘


Nie jest perfekcyjnie, bo zostaje zamknięcie cudzysłowiu i dwukropek, ale nie będzie generować błędów.

Wracając do meritum, jeżeli zrobię klasyfikację tylko czy dany tag jest pierwszym tagiem w tej grze oszczędzi mi ok 7% wymiarów, natomiast zrobienie oddzielnie tagu 1 od 2 zwiększa mi już drastycznie liczbę wymiarów. Więc to nie ma sensu. Można zrobić po prostu klasyfikację czy jest ten tag czy nie i zachować 452 wymiary. Szczególnie, że ułatwi to później dzielenie zbioru danych na treningowy i testowy przez zmniejszenie ryzyka braku konkretnego tagu w danych treningowych. Kusi mnie zastosować kryterium, że musi być minimum x głosów za jakimś tagiem, żeby wyeliminować losowe kliknięcia przydzielające tag.

### Kolumny liczbowe

Ze względu na gigantyczną liczbę kolumn, trzeba je jakoś masowo posprawdzać czy mają sens i potrzebę istnienia. Jednak zanim przejdziemy do czystek, sprawdźmy jakie kolumny są liczbowe:

In [23]:
import polars.selectors as cs

numeric_cols = df.select(cs.numeric()).describe()
print("Kolumny liczbowe:")
print(numeric_cols)

Kolumny liczbowe:
shape: (9, 19)
┌────────────┬──────────────┬──────────┬───────────┬──────────────────┬──────────────┬─────────────────┬──────────┬────────────┬────────────┬──────────────┬────────────┬──────────────────────────┬─────────────────────────┬─────────────────────────┬────────────────────────┬───────────┬─────────────┬───────────────────┐
│ statistic  ┆ required_age ┆ price    ┆ dlc_count ┆ metacritic_score ┆ achievements ┆ recommendations ┆ movies   ┆ user_score ┆ score_rank ┆ positive     ┆ negative   ┆ average_playtime_forever ┆ average_playtime_2weeks ┆ median_playtime_forever ┆ median_playtime_2weeks ┆ discount  ┆ peak_ccu    ┆ __index_level_0__ │
│ ---        ┆ ---          ┆ ---      ┆ ---       ┆ ---              ┆ ---          ┆ ---             ┆ ---      ┆ ---        ┆ ---        ┆ ---          ┆ ---        ┆ ---                      ┆ ---                     ┆ ---                     ┆ ---                    ┆ ---       ┆ ---         ┆ ---               │
│ str  

Od razu rzuca się w oczy fakt, że kolumna `movies` jest cała pusta. Podejrzewam, że twórca datasetu miał jakiś błąd podczas scrapowania, więc tą kolumnę trzeba usunąć. Kolumna `score_rank` również jest prawie całkowicie pusta. Ma jedynie 40 rekordów niepustych, choć to akurat w wyniku działania samej platformy steam, ponieważ liczy tą metrykę tylko dla najpopularniejszych gier. Oznacza to, że muszę również tą kolumnę usunąć do kolejnych działań, żeby później nie zepsuć sobie modeli predykcji czy jakaś gra będzie hitem. Model szybko by się zbiasował i byłby gorszy.

In [24]:
df = df.drop(["movies", "score_rank"])

Jedną z metod stosowanych przez ekonometryków jest sprawdzanie korelacji pomiędzy kolumnami. Jeżeli dwie kolumny mają podobne zależności, czyli rosną w podobnym stopniu przy zmianach innych parametrów i też spadają podobnie, to można usunąć jedną cechę z tej pary, ponieważ przyjmuje się, że jedna z tych cech jest dobrym estymatorem drugiej cechy.

Z doświadczenia używam w tym celu korelacji liniowej pearsona lub korelacji rangowej spearmana. Niestety ta pierwsza jest nieodporna na wartości odstające, a ta druga traci cenną informację o dystansie między próbkami (o ile dystans jest policzalny).